istio-client query virtualservice get error: Could not resolve subtype of [simple type, class io.fabric8.istio.api.networking.v1beta1.IsHTTPRedirectRedirectPort]:

### Describe the bug

```
  - match:
    - uri:
        exact: /
    redirect:
      uri: /login
```

When virtualService has HTTPRedirect ,use istio-client query virtualservice list get a error:

Caused by: com.fasterxml.jackson.databind.exc.InvalidTypeIdException: Could not resolve subtype of [simple type, class io.fabric8.istio.api.networking.v1beta1.IsHTTPRedirectRedirectPort]: Cannot deduce unique subtype of `io.fabric8.istio.api.networking.v1beta1.IsHTTPRedirectRedirectPort` (2 candidates match)\n at [Source: UNKNOWN; line: -1, column: -1] (through reference chain: io.fabric8.istio.api.networking.v1beta1.VirtualService[\"spec\"]->io.fabric8.istio.api.networking.v1beta1.VirtualServiceSpec[\"http\"]->java.util.ArrayList[1]->io.fabric8.istio.api.networking.v1beta1.HTTPRoute[\"redirect\"]->io.fabric8.istio.api.networking.v1beta1.HTTPRedirect[\"RedirectPort\"])\n\tat com.fasterxml.jackson.databind.exc.InvalidTypeIdException.from(InvalidTypeIdException.java:43)\n\tat 

### Fabric8 Kubernetes Client version

5.12.2

### Steps to reproduce

When virtualService has HTTPRedirect,yaml like this:
```
  - match:
    - uri:
        exact: /
    redirect:
      uri: /login
```

use istio-client query virtualservice list get a error:

```
Caused by: com.fasterxml.jackson.databind.exc.InvalidTypeIdException: Could not resolve subtype of [simple type, class io.fabric8.istio.api.networking.v1beta1.IsHTTPRedirectRedirectPort]: Cannot deduce unique subtype of `io.fabric8.istio.api.networking.v1beta1.IsHTTPRedirectRedirectPort` (2 candidates match)\n at [Source: UNKNOWN; line: -1, column: -1] (through reference chain: io.fabric8.istio.api.networking.v1beta1.VirtualService[\"spec\"]->io.fabric8.istio.api.networking.v1beta1.VirtualServiceSpec[\"http\"]->java.util.ArrayList[1]->io.fabric8.istio.api.networking.v1beta1.HTTPRoute[\"redirect\"]->io.fabric8.istio.api.networking.v1beta1.HTTPRedirect[\"RedirectPort\"])\n\tat com.fasterxml.jackson.databind.exc.InvalidTypeIdException.from(InvalidTypeIdException.java:43)\n\tat 

com.fasterxml.jackson.databind.DeserializationContext.missingTypeIdException(DeserializationContext.java:1943)\n\tat com.fasterxml.jackson.databind.DeserializationContext.handleMissingTypeId(DeserializationContext.java:1456)\n\tat com.fasterxml.jackson.databind.jsontype.impl.TypeDeserializerBase._handleMissingTypeId(TypeDeserializerBase.java:307)\n\tat com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedUsingDefaultImpl(AsPropertyTypeDeserializer.java:178)\n\tat com.fasterxml.jackson.databind.jsontype.impl.AsDeductionTypeDeserializer.deserializeTypedFromObject(AsDeductionTypeDeserializer.java:133)\n\tat com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:263)\n\tat com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:138)\n\tat io.fabric8.kubernetes.client.utils.serialization.SettableBeanPropertyDelegate.deserializeAndSet(SettableBeanPropertyDelegate.java:131)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:324)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:187)\n\tat io.fabric8.kubernetes.model.jackson.JsonUnwrappedDeserializer.deserialize(JsonUnwrappedDeserializer.java:139)\n\tat com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)\n\tat io.fabric8.kubernetes.client.utils.serialization.SettableBeanPropertyDelegate.deserializeAndSet(SettableBeanPropertyDelegate.java:131)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:324)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:187)\n\tat com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:347)\n\tat com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:244)\n\tat com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:28)\n\tat com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)\n\tat io.fabric8.kubernetes.client.utils.serialization.SettableBeanPropertyDelegate.deserializeAndSet(SettableBeanPropertyDelegate.java:131)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:324)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:187)\n\tat com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)\n\tat io.fabric8.kubernetes.client.utils.serialization.SettableBeanPropertyDelegate.deserializeAndSet(SettableBeanPropertyDelegate.java:131)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:324)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:187)\n\tat com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:322)\n\tat com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:4569)\n\tat com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2798)\n\tat com.fasterxml.jackson.databind.ObjectMapper.treeToValue(ObjectMapper.java:3261)\n\tat io.fabric8.kubernetes.internal.KubernetesDeserializer.fromObjectNode(KubernetesDeserializer.java:99)\n\tat io.fabric8.kubernetes.internal.KubernetesDeserializer.deserialize(KubernetesDeserializer.java:63)\n\tat io.fabric8.kubernetes.internal.KubernetesDeserializer.deserialize(KubernetesDeserializer.java:45)\n\tat com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:322)\n\tat com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:2033)\n\tat com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1491)\n\tat io.fabric8.kubernetes.client.utils.Serialization.unmarshalJsonStr(Serialization.java:383)\n\tat io.fabric8.kubernetes.client.utils.Serialization.unmarshalYaml(Serialization.java:376)\n\tat io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:196)\n\t... 140 more\n"}
```

### Expected behavior

success

### Runtime

Kubernetes (vanilla)

### Kubernetes API Server version

1.22.3@latest

### Environment

Linux

### Fabric8 Kubernetes Client Logs

_No response_

### Additional context

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@
 #### Improvements
 * Fix #4041: adding Quantity.getNumericalAmount with an explanation about bytes and cores.
 * Fix #4241: added more context to informer logs with the endpoint path
+* Fix #4250: allowing for deserialization of polymorphic unwrapped fields
 
 #### Dependency Upgrade
 

--- a/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/Serialization.java
+++ b/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/Serialization.java
@@ -17,6 +17,7 @@ package io.fabric8.kubernetes.client.utils;
 
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.DeserializationFeature;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
 import com.fasterxml.jackson.dataformat.yaml.YAMLGenerator;
@@ -52,6 +53,7 @@ public class Serialization {
   private static final ObjectMapper JSON_MAPPER = new ObjectMapper();
   static {
     JSON_MAPPER.registerModules(new JavaTimeModule(), UNMATCHED_FIELD_TYPE_MODULE);
+    JSON_MAPPER.disable(DeserializationFeature.FAIL_ON_INVALID_SUBTYPE);
   }
 
   private static volatile ObjectMapper YAML_MAPPER;

--- a/kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/utils/SerializationTest.java
+++ b/kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/utils/SerializationTest.java
@@ -15,8 +15,12 @@
  */
 package io.fabric8.kubernetes.client.utils;
 
+import com.fasterxml.jackson.annotation.JsonInclude;
 import com.fasterxml.jackson.annotation.JsonSubTypes;
 import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.annotation.JsonUnwrapped;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonTypeResolver;
 import com.fasterxml.jackson.databind.node.BooleanNode;
 import com.fasterxml.jackson.databind.node.IntNode;
 import com.fasterxml.jackson.databind.node.TextNode;
@@ -50,6 +54,8 @@ import java.util.Map;
 import java.util.stream.Collectors;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertNull;
+import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
 class SerializationTest {
@@ -320,6 +326,46 @@ class SerializationTest {
   void quantityQuoting() {
     Quantity quantity = Serialization.unmarshal("amount: \"2\"\nformat: \"Gi\"", Quantity.class);
     assertThat(Serialization.asYaml(quantity)).isEqualTo("--- \"2Gi\"\n");
+  }
+
+  @JsonTypeResolver(io.fabric8.kubernetes.model.jackson.UnwrappedTypeResolverBuilder.class)
+  @JsonInclude(JsonInclude.Include.NON_NULL)
+  @JsonSubTypes({
+      @JsonSubTypes.Type(Named.class),
+      @JsonSubTypes.Type(Counted.class)
+  })
+  @JsonTypeInfo(use = JsonTypeInfo.Id.DEDUCTION)
+  interface Poly {
+
+  }
+
+  @JsonDeserialize(using = com.fasterxml.jackson.databind.JsonDeserializer.None.class)
+  @JsonInclude(JsonInclude.Include.NON_NULL)
+  static class Named implements Poly {
+    public String name;
+  }
+
+  @JsonDeserialize(using = com.fasterxml.jackson.databind.JsonDeserializer.None.class)
+  static class Counted implements Poly {
+    public int count;
+  }
+
+  @JsonDeserialize(using = io.fabric8.kubernetes.model.jackson.JsonUnwrappedDeserializer.class)
+  @JsonInclude(JsonInclude.Include.NON_NULL)
+  static class PolyParent {
+    @JsonUnwrapped
+    public Poly poly;
+    public int other;
+  }
+
+  @Test
+  void polymorphicUnwrap() {
+    PolyParent parent = Serialization.unmarshal("other: 1\nname: x", PolyParent.class);
+    assertEquals("x", ((Named) parent.poly).name);
+
+    // should be null if it can't be deduced
+    parent = Serialization.unmarshal("other: 1", PolyParent.class);
+    assertNull(parent.poly);
   }
 
 }

